src/NServiceBus.Hosting.Tests/StartablesAndStoppables/StartableAndStoppableRunnerTests.cs

```diff
@@ -0,0 +1,283 @@
+            {
+                Started = true;
+                return TaskEx.CompletedTask;
+            }
+
+            public Task Stop(IMessageSession session)
+            {
+                Stopped = true;
+                return TaskEx.CompletedTask;
+            }
+        }
+
+        class StartableStartReturnsNull : IWantToRunWhenEndpointStartsAndStops
+        {
+            public Task Start(IMessageSession session)
+            {
+                return null;
+            }
+
+            public Task Stop(IMessageSession session)
+            {
+                throw new NotImplementedException();
+            }
+        }
+
+        class StartableStopReturnsNull : IWantToRunWhenEndpointStartsAndStops
+        {
+            public Task Start(IMessageSession session)
+            {
+                throw new NotImplementedException();
+            }
+
+            public Task Stop(IMessageSession session)
+            {
+                return null;
+            }
+        }
+
+        class SyncThrowingStart : IWantToRunWhenEndpointStartsAndStops
+        {
+            public bool Stopped { get; set; }
+
+            public Task Start(IMessageSession session)
+            {
+                throw new InvalidOperationException("SyncThrowingStart");
+            }
+
+            public Task Stop(IMessageSession session)
+            {
+                Stopped = true;
+                return TaskEx.CompletedTask;
+            }
+        }
+
+        class AsyncThrowingStart : IWantToRunWhenEndpointStartsAndStops
+        {
+            public bool Stopped { get; set; }
+
+            public async Task Start(IMessageSession session)
+            {
+                await Task.Yield();
+                throw new InvalidOperationException("AsyncThrowingStart");
+            }
+
+            public Task Stop(IMessageSession session)
+            {
+                Stopped = true;
+                return TaskEx.CompletedTask;
+            }
+        }
+
+        class SyncThrowingStop : IWantToRunWhenEndpointStartsAndStops
+        {
+            public bool Started { get; set; }
+
+            public Task Start(IMessageSession session)
+            {
+                Started = true;
+                return TaskEx.CompletedTask;
+            }
+
+            public Task Stop(IMessageSession session)
+            {
+                throw new InvalidOperationException("SyncThrowingStop");
+            }
+        }
+
+        class AsyncThrowingStop : IWantToRunWhenEndpointStartsAndStops
+        {
+            public bool Started { get; set; }
+
+            public Task Start(IMessageSession session)
+            {
+                Started = true;
+                return TaskEx.CompletedTask;
+            }
+
+            public async Task Stop(IMessageSession session)
+            {
+                await Task.Yield();
+                throw new InvalidOperationException("AsyncThrowingStop");
+            }
+        }
+    }
+}
```

Comment:
I think it would make sense to prefix the test names with the Method you're unit testing. (e.g. `Start_Should[...]`)